### PR TITLE
Refactor snapshot build

### PR DESCRIPTION
**Pull request by ry**
Depends on https://github.com/denoland/deno/pull/2827

Fixes https://github.com/denoland/deno/issues/2496
Fixes https://github.com/denoland/deno/issues/2711
Fixes https://github.com/denoland/deno/issues/459
Towards https://github.com/denoland/deno/issues/2608
Towards https://github.com/denoland/deno/issues/986

Prototype: https://github.com/ry/deno_typescript

|   | before (master, 65fa2b8)  | after |
| ------------- | ------------- | ------------- |
| build.py deno; from scratch | 2m46.347s  | 2m28.869s  |
| cargo build; from scratch | 6m 35s | 4m 14s  |
| build.py deno; touch js/util.ts | 1m25.620s  | 0m24.117s  |
| cargo build; touch js/util.ts | BROKEN  | 43.68s  |

This patch removes the build dependency on Node and rollup and consequently deletes approximately 1 million lines from third_party/node_modules.

A Rust Crate was published with public V8 snapshotting API  https://crates.io/crates/deno_typescript

### CODE REVIEW

File: cli/build.rs [hunk 1]
```diff
@@ -1,15 +0,0 @@
...
-
-fn main() {
-  let build = gn::Build::setup();
-  // When RLS is running "cargo check" to analyze the source code, we're not
-  // trying to build a working executable, rather we're just compiling all
-  // rust code.
-  if !build.check_only {
-    build.run("cli:deno_deps");
-  }
-}
```
Comment (ry) [inline cli/build.rs]: Note that cli/build.rs is removed. That means it does not depend on GN in "cargo build"

File: cli/BUILD.gn [hunk 2]
```diff
@@ -189,41 +85,10 @@ rust_test("cli_test") {
...
-  ]
-}
-
-# Generates $target_gen_dir/snapshot_compiler.bin
-snapshot("snapshot_compiler") {
-  source_root = "$target_gen_dir/bundle/compiler.js"
-  deps = [
-    ":compiler_bundle",
-  ]
-}
```
Comment (ry) [inline cli/BUILD.gn]: Note existing snapshot infrastructure is removed.

File: rollup.config.js [hunk 3]
```diff
@@ -1,231 +0,0 @@
-// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
```
Comment (ry) [inline rollup.config.js]: Note that rollup is removed.

File: deno_typescript/lib.rs [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+  let main = state.written_files.last().unwrap().module_name.clone();
+  // let main = "file:///Users/rld/src/deno_typescript/cli_snapshots/js/main.ts";
+  js_check(runtime_isolate.execute("anon", &format!("require('{}')", main)));
+
+  snapshot_to_env(runtime_isolate, env_var)?;
+
+  Ok(())
+}
+
+pub fn mksnapshot_bundle_ts(
```
Comment (afinch7) [inline deno_typescript/lib.rs]: Maybe add a triple slash doc, so we know that `mksnapshot_bundle_ts` includes typescript in the snapshot.

File: deno_typescript/lib.rs [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    runtime_isolate.execute("amd_runtime.js", include_str!("amd_runtime.js")),
+  );
+
+  js_check(runtime_isolate.execute(&bundle.to_string_lossy(), &source_code));
+
+  // execute main.
+
+  let state = state.lock().unwrap();
+  let main = state.written_files.last().unwrap().module_name.clone();
+  // let main = "file:///Users/rld/src/deno_typescript/cli_snapshots/js/main.ts";
```
Comment (afinch7) [inline deno_typescript/lib.rs]: rm?

File: deno_typescript/lib.rs [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+  js_check(runtime_isolate.execute("anon", &format!("require('{}')", main)));
+
+  snapshot_to_env(runtime_isolate, env_var)?;
+
+  Ok(())
+}
+
+// TODO(ry) Instead of state: Arc<Mutex<TSState>>, take something like state:
+// &TSState
+pub fn mksnapshot(
```
Comment (afinch7) [inline deno_typescript/lib.rs]: What does this do?

File: cli_snapshots/print_hello.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export function printHello(): void {
```
Comment (afinch7) [inline cli_snapshots/print_hello.ts]: Is this file still needed?

Comment (ry) [reply cli_snapshots/print_hello.ts]: Removed.

File: deno_typescript/lib.rs (hunk 6 above)
Comment (ry) [reply deno_typescript/lib.rs]: Removed. Dead code.

File: deno_typescript/lib.rs (hunk 5 above)
Comment (ry) [reply deno_typescript/lib.rs]: Done

File: deno_typescript/lib.rs (hunk 4 above)
Comment (ry) [reply deno_typescript/lib.rs]: Done

File: deno_typescript/compiler_main.js [hunk 8]
```diff
@@ -0,0 +1,332 @@
+// Because we're bootstrapping the TS compiler without dependencies on Node,
```
Comment (piscisaureus) [inline deno_typescript/compiler_main.js]: Nice that I get to read this file now.

File: cli_snapshots/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+edition = "2018"
+description = "Provides snapshots for the deno CLI"
+repository = "https://github.com/ry/deno_typescript"
+readme = "../README.md"
+
+[lib]
+path = "lib.rs"
+
+[dev-dependencies]
+deno = "0.15.0"
```
Comment (afinch7) [inline cli_snapshots/Cargo.toml]: `path = "../core"`?

Comment (ry) [reply cli_snapshots/Cargo.toml]: Good catch! Fixed.

**Comment by ry** [comment]
@afinch7 Thanks for the review. Yes, the full path of where the executable is built is embedded in the source maps. I will address that after this lands.

File: deno_typescript/lib.rs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+  js_check(
+    runtime_isolate.execute("amd_runtime.js", include_str!("amd_runtime.js")),
+  );
+
+  js_check(runtime_isolate.execute(&bundle.to_string_lossy(), &source_code));
+
+  // execute main.
+
+  let state = state.lock().unwrap();
+  let main = state.written_files.last().unwrap().module_name.clone();
```
Comment (bartlomieju) [inline deno_typescript/lib.rs]: Can this be abstracted as method on `TsState` with short comment? It's not obvious that last written file is first root name.

File: deno_typescript/compiler_main.js [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+    println(msg);
+    if (rest) {
+      println(`And ${rest} other errors.`);
+    }
+    exit(1);
+  }
+}
+
+/** Returns the TypeScript Extension enum for a given media type. */
+function getExtension(fileName) {
```
Comment (bartlomieju) [inline deno_typescript/compiler_main.js]: This is interesting - on one hand I think this method won't work with extensionless files that were given content type in headers but on the other it allows to implement background download of multiple modules.

File: js/version.ts [hunk 12]
```diff
@@ -9,7 +9,7 @@ export const version: Version = {
   deno: "",
   v8: "",
   // This string will be replaced by rollup
-  typescript: `ROLLUP_REPLACE_TS_VERSION`
+  typescript: `DENO_REPLACE_TS_VERSION`
```
Comment (bartlomieju) [inline js/version.ts]: If this value is injected via Rust then `Version` should be removed and `deno --version` should be handled only in Rust (of course in followup PR)

CC @nayeemrmn

Comment (ry) [reply js/version.ts]: It’s actually injected in JS...

File: .appveyor.yml [hunk 13]
```diff
@@ -216,35 +216,15 @@ after_test:
       }
 
   # Verify that javascript and typescript files which are bundled by rollup are
```
[the remark this answers is not in the document]
Comment (ry) [reply .appveyor.yml]: Fixed

File: deno_typescript/lib.rs [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+  let mut replacements = HashMap::new();
+  replacements.insert("DENO_REPLACE_OS", BUILD_OS);
+  replacements.insert("DENO_REPLACE_ARCH", BUILD_ARCH);
+  serde_json::json!(replacements).to_string()
+}
+
+impl TSIsolate {
+  fn new(bundle: bool) -> TSIsolate {
+    let mut isolate = Isolate::new(StartupData::None, false);
+    let main_code = include_str!("compiler_main.js");
```
[the remark this answers is not in the document]
Comment (ry) [reply deno_typescript/lib.rs]: Done

File: deno_typescript/lib.rs [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+  bundle: &Path,
+  env_var: &str,
+  state: Arc<Mutex<TSState>>,
+) -> Result<(), ErrBox> {
+  let mut runtime_isolate = Isolate::new(StartupData::None, true);
+  let source_code_vec = std::fs::read(bundle)?;
+  let source_code = std::str::from_utf8(&source_code_vec)?;
+
+  js_check(
+    runtime_isolate.execute("amd_runtime.js", include_str!("amd_runtime.js")),
```
[the remark this answers is not in the document]
Comment (ry) [reply deno_typescript/lib.rs]: Done

File: deno_typescript/lib.rs [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+  });
+
+  let mut root_names_str: Vec<String> = root_names
+    .iter()
+    .map(|p| {
+      if !p.exists() {
+        panic!("File not found {}", p.display());
+      }
+
+      let module_specifier =
```
[the remark this answers is not in the document]
Comment (ry) [reply deno_typescript/lib.rs]: deno_typescript::compile was dead code from a previous experiments. I've removed it now.

File: deno_typescript/lib.rs (hunk 10 above)
Comment (ry) [reply deno_typescript/lib.rs]: Yes this is quite hacky and needs a comment. Fixed.

File: deno_typescript/compiler_main.js (hunk 11 above)
Comment (ry) [reply deno_typescript/compiler_main.js]: Yes, we have two compilers now. I would love to merge them, but that is going to take further work. This is just the first step.

Comment (bartlomieju) [reply deno_typescript/compiler_main.js]: How about bootstrapping? Theoretically we could make Deno to bundle code in `js/`...

Comment (ry) [reply deno_typescript/compiler_main.js]: No, I think that introduces too much complexity. I don't want our build to depend on binaries.

Comment (ry) [reply deno_typescript/compiler_main.js]: My plan is to slowly port js/compiler.ts to javascript and move it into deno_typescript.

File: build_extra/rust/run.py [hunk 17]
```diff
@@ -28,6 +28,7 @@
 # When building with Cargo this variable is set by build.rs.
 env["GN_OUT_DIR"] = os.path.abspath(".")
 assert os.path.isdir(env["GN_OUT_DIR"])
+env["OUT_DIR"] = os.path.abspath(".")
```
Comment (piscisaureus) [inline build_extra/rust/run.py]: Add it to the appropriate GN target: `env += [ "OUT_DIR=." ]` or use --generated-source-dir.

File: cli_snapshots/build.rs [hunk 18]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+use std::env;
+use std::path::PathBuf;
+
+fn main() {
+  // deno_typescript::trace_serializer();
+
+  let c = PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").unwrap());
```
Comment (piscisaureus) [inline cli_snapshots/build.rs]: Where does this get set when building with GN?
Or does this not run during the GN build?

Comment (piscisaureus) [reply cli_snapshots/build.rs]: Oh ok nvm - it has its own run.py

Comment (ry) [reply cli_snapshots/build.rs]: This is run with GN, it's set in `cli_snapshots/run.py`.

File: deno_typescript/amd_runtime.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+  function createOrLoadModule(name) {
+    if (!modules.has(name)) {
+      const m = { name, exports: {} };
+      modules.set(name, m);
+    }
+    return modules.get(name);
+  }
+
+  require = name => {
+    // println("globalRequire");
```
Comment (piscisaureus) [inline deno_typescript/amd_runtime.js]: Remove

File: deno_typescript/amd_runtime.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    return createOrLoadModule(name).exports;
+  };
+
+  define = (name, deps, factory) => {
+    const currentModule = createOrLoadModule(name);
+    const localExports = currentModule.exports;
+    // println("localDefine", name, deps, localExports);
```
Comment (piscisaureus) [inline deno_typescript/amd_runtime.js]: Ditto

File: deno_typescript/compiler_main.js [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+
+  const program = ts.createProgram(rootNames, options, host);
+
+  diagnostics = ts.getPreEmitDiagnostics(program).filter(({ code }) => {
+    // TS2691: An import path cannot end with a '.ts' extension. Consider
+    // importing 'bad-module' instead.
+    if (code === 2691) return false;
+    // TS5009: Cannot find the common subdirectory path for the input files.
+    if (code === 5009) return false;
+    /*
```
Comment (piscisaureus) [inline deno_typescript/compiler_main.js]: Why are these commented out?
Can we drop them altogether?

File: deno_typescript/compiler_main.js [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+  for (let i = 0; i < ui8.length; i++) {
+    out += String.fromCharCode(ui8[i]);
+  }
+  return out;
+}
+
+function encode(str) {
+  const charCodes = str.split("").map(c => c.charCodeAt(0));
+  const ui8 = new Uint8Array(charCodes);
+  // println(`encode ${ui8}`);
```
Comment (piscisaureus) [inline deno_typescript/compiler_main.js]: Renove

File: deno_typescript/compiler_main.js [hunk 23]
```diff
@@ -0,0 +1,332 @@
...
+  setEmitResult: 53
+};
+
+// interface CompilerHost extends ModuleResolutionHost {
+class Host {
+  // fileExists(fileName: string): boolean;
+  fileExists(fileName) {
+    println(`fileExists ${fileName}`);
+    return true;
+    // unreachable();
```
Comment (piscisaureus) [inline deno_typescript/compiler_main.js]: Remove

File: deno_typescript/compiler_main.js [hunk 24]
```diff
@@ -0,0 +1,332 @@
...
+  writeFile: 51,
+  resolveModuleNames: 52,
+  setEmitResult: 53
+};
+
+// interface CompilerHost extends ModuleResolutionHost {
+class Host {
+  // fileExists(fileName: string): boolean;
+  fileExists(fileName) {
+    println(`fileExists ${fileName}`);
```
Comment (piscisaureus) [inline deno_typescript/compiler_main.js]: Remove

File: deno_typescript/compiler_main.js [hunk 25]
```diff
@@ -0,0 +1,332 @@
...
+        .replace("/index.d.ts", "");
+    }
+
+    let { sourceCode, moduleName } = dispatch("readFile", {
+      fileName,
+      languageVersion,
+      shouldCreateNewSourceFile
+    });
+
+    // TODO(ry) A terrible hack. Please remove ASAP.
```
Comment (piscisaureus) [inline deno_typescript/compiler_main.js]: Huu

File: deno_typescript/lib.deno_core.d.ts [hunk 26]
```diff
@@ -1,5 +1,13 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
 
+// This file contains APIs that are introduced into the global namespace by
```
Comment (piscisaureus) [inline deno_typescript/lib.deno_core.d.ts]: I'd say that from the CLI user perspective we should place these APIS under Deno.internal.xyz rather than under Deno.core. The former suggests "these are built in functions", the latter tells the user "Unstable internal interface, do not use ".

File: js/main.ts [hunk 27]
```diff
@@ -1,36 +1,29 @@
...
   const s = os.start(preserveDenoNamespace, name);
 
   setVersions(s.denoVersion, s.v8Version);
 
   // handle `--version`
   if (s.versionFlag) {
-    console.log("deno:", deno.version.deno);
-    console.log("v8:", deno.version.v8);
-    console.log("typescript:", deno.version.typescript);
+    const { console } = window;
```
Comment (piscisaureus) [inline js/main.ts]: Unnecessary?

File: deno_typescript/lib.deno_core.d.ts (hunk 26 above)
Comment (ry) [reply deno_typescript/lib.deno_core.d.ts]: Let's discuss that in an issue, it's unrelated to this PR.

File: js/repl.ts [hunk 28]
```diff
@@ -1,12 +1,14 @@
...
+import { close } from "./files.ts";
+import { exit } from "./os.ts";
+import { window } from "./window.ts";
+import { core } from "./core.ts";
+import { formatError } from "./format_error.ts";
+import { stringifyArgs } from "./console.ts";
+import * as dispatch from "./dispatch.ts";
+import { sendSync, sendAsync } from "./dispatch_json.ts";
+
+const { console } = window;
```
Comment (piscisaureus) [inline js/repl.ts]: I don't understand. Is window no longer the global object?

File: js/timers.ts [hunk 29]
```diff
@@ -1,8 +1,10 @@
...
-import { assert } from "./util";
-import { window } from "./window";
-import * as dispatch from "./dispatch";
-import { sendSync, sendAsync } from "./dispatch_json";
+import { assert } from "./util.ts";
+import { window } from "./window.ts";
+import * as dispatch from "./dispatch.ts";
+import { sendSync, sendAsync } from "./dispatch_json.ts";
+
+const { console } = window;
```
Comment (piscisaureus) [inline js/timers.ts]: ^

File: js/util.ts [hunk 30]
```diff
@@ -1,5 +1,7 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
-import { TypedArray } from "./types";
+import { TypedArray } from "./types.ts";
+import { window } from "./window.ts";
+const { console } = window;
```
Comment (piscisaureus) [inline js/util.ts]: ^

File: build_extra/rust/run.py (hunk 17 above)
Comment (ry) [reply build_extra/rust/run.py]: done

File: deno_typescript/ops.rs [hunk 31]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+fn dispatch2(
+  s: &mut TSState,
+  op_id: OpId,
+  control_buf: &[u8],
+) -> Result<Value, ErrBox> {
+  let v = serde_json::from_slice(control_buf)?;
+  match op_id {
+    49 => read_file(s, v),
```
Comment (piscisaureus) [inline deno_typescript/ops.rs]: Add a comment that these need to be kept in sync with the other places OP IDs are hard coded.
(Maybe add a mod op_ids so they are hard coded only once in rust?)

File: deno_typescript/ops.rs [hunk 32]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+pub fn dispatch_op(s: &mut TSState, op_id: OpId, control_buf: &[u8]) -> CoreOp {
+  let result = dispatch2(s, op_id, control_buf);
+  let response = match result {
+    Ok(v) => json!({ "ok": v }),
+    Err(err) => json!({ "err": err.to_string() }),
+  };
+  let x = serde_json::to_string(&response).unwrap();
+  let vec = x.to_string().into_bytes();
```
Comment (piscisaureus) [inline deno_typescript/ops.rs]: Unnecessary to_string()

File: js/main.ts (hunk 27 above)
Comment (ry) [reply js/main.ts]: Necessary - here and elsewhere. Somehow the ambient types changed - I think maybe because we no longer have `libs: ["dom"]` in the tsconfig... I'm not sure.

File: deno_typescript/ops.rs [hunk 33]
```diff
@@ -0,0 +1,160 @@
...
+  file_name: String,
+  language_version: Option<i32>,
+  should_create_new_source_file: bool,
+}
+
+fn read_file(_s: &mut TSState, v: Value) -> Result<Value, ErrBox> {
+  let v: ReadFile = serde_json::from_value(v)?;
+  let (module_name, source_code) = if v.file_name.starts_with("$asset$/") {
+    let asset = v.file_name.replace("$asset$/", "");
+    // TODO(ry) remove unwrap below, return error "no such asset"
```
Comment (piscisaureus) [inline deno_typescript/ops.rs]: io::Error::new(io::ErrorKind::NotFound, "Asset not found");

File: deno_typescript/ops.rs [hunk 34]
```diff
@@ -0,0 +1,160 @@
...
+struct ResolveModuleNames {
+  module_names: Vec<String>,
+  containing_file: String,
+}
+
+fn resolve_module_names(_s: &mut TSState, v: Value) -> Result<Value, ErrBox> {
+  let v: ResolveModuleNames = serde_json::from_value(v).unwrap();
+  let mut resolved = Vec::<String>::new();
+
+  /*
```
Comment (piscisaureus) [inline deno_typescript/ops.rs]: Remove?

File: deno_typescript/ops.rs [hunk 35]
```diff
@@ -0,0 +1,160 @@
...
+  s.exit_code = v.code;
+  std::process::exit(v.code)
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct EmitResult {
+  pub emit_skipped: bool,
+  pub diagnostics: Vec<String>,
+  pub emitted_files: Vec<String>, // "sourceMaps": []
```
Comment (piscisaureus) [inline deno_typescript/ops.rs]: Weird comment

File: deno_typescript/ops.rs (hunk 31 above)
Comment (ry) [reply deno_typescript/ops.rs]: > Add a comment that these need to be kept in sync with the other places OP IDs are hard coded.

Done.

> (Maybe add a mod op_ids so they are hard coded only once in rust?)

I think we need a better solution that doesn't have any runtime overhead, which can also be used in the CLI. I like the `derive(TypeScript)` idea... but yea. That's for a different PR.

File: deno_typescript/lib.rs [hunk 36]
```diff
@@ -0,0 +1,278 @@
...
+        ModuleSpecifier::resolve_url_or_path(&p.to_string_lossy()).unwrap();
+      module_specifier.as_str().to_string()
+      // p.to_string_lossy().to_string()
```
Comment (piscisaureus) [inline deno_typescript/lib.rs]: Remove

File: deno_typescript/ops.rs (hunk 32 above)
Comment (ry) [reply deno_typescript/ops.rs]: Done

File: deno_typescript/ops.rs (hunk 33 above)
Comment (ry) [reply deno_typescript/ops.rs]: Done

File: deno_typescript/ops.rs (hunk 34 above)
Comment (ry) [reply deno_typescript/ops.rs]: Done

File: deno_typescript/ops.rs (hunk 35 above)
Comment (ry) [reply deno_typescript/ops.rs]: Removed.

File: deno_typescript/lib.rs (hunk 36 above)
Comment (ry) [reply deno_typescript/lib.rs]: Done

**Comment by ry** [comment]
This patch reduced executable size by 7mb. It had some effect on some of the benchmark's memory usage as well.

Screen Shot 2019-09-02 at 6 11 56 PM

File: js/build.ts [hunk 37]
```diff
@@ -18,9 +18,9 @@ export interface BuildInfo {
 export const build: BuildInfo = {
   // These string will be replaced by rollup
```
Comment (hexagon6) [inline js/build.ts]: comment about rollup will be outdated.
